### PR TITLE
feat: force kill daemons after timeout

### PR DESCRIPTION
**Pull request by alanshaw**
This PR re-adds a feature that was removed in 1.0.0 whereby daemons are force killed with a SIGKILL when they do not stop within a timeout.

It adds 2 new options to the `Daemon` class - `forceKill` (boolean - default `true`) and `forceKillTimeout` (number - default 10000).

resolves https://github.com/ipfs/js-ipfsd-ctl/issues/438

### CODE REVIEW

File: src/ipfsd-daemon.js [hunk 1]
```diff
@@ -239,7 +239,24 @@ class Daemon {
       return this
     }
 
-    await this.api.stop()
+    let killTimeout
+    let killed = false
+    if (this.opts.forceKill !== false) {
```
Comment (hugomrdias) [inline src/ipfsd-daemon.js]: can you add default to true here https://github.com/ipfs/js-ipfsd-ctl/blob/3dae27670fdebe0edd6e1e26ddf5df4c274b62bf/src/factory.js#L20 and in the docs

File: src/ipfsd-daemon.js [hunk 2]
```diff
@@ -239,7 +239,24 @@ class Daemon {
...
+      killTimeout = setTimeout(() => {
+        // eslint-disable-next-line no-console
+        console.error(new Error(`Timeout stopping ${this.opts.type} node. Process ${this.subprocess.pid} will be force killed now.`))
+        killed = true
+        this.subprocess.kill('SIGKILL')
+      }, this.opts.forceKillTimeout || 10000)
```
Comment (hugomrdias) [inline src/ipfsd-daemon.js]: can you add a default to 10000 here https://github.com/ipfs/js-ipfsd-ctl/blob/3dae27670fdebe0edd6e1e26ddf5df4c274b62bf/src/factory.js#L20 and in the docs

File: src/ipfsd-daemon.js [hunk 3]
```diff
@@ -239,7 +239,24 @@ class Daemon {
...
+    }
+
+    try {
+      await this.api.stop()
+    } catch (err) {
+      if (!killed) throw err // if was killed then ignore error
```
Comment (hugomrdias) [inline src/ipfsd-daemon.js]: can you add a debug for the err when killed pls

File: src/ipfsd-daemon.js [hunk 4]
```diff
@@ -239,7 +239,24 @@ class Daemon {
...
 
-    await this.api.stop()
+    let killTimeout
+    let killed = false
+    if (this.opts.forceKill !== false) {
+      killTimeout = setTimeout(() => {
+        // eslint-disable-next-line no-console
+        console.error(new Error(`Timeout stopping ${this.opts.type} node. Process ${this.subprocess.pid} will be force killed now.`))
+        killed = true
+        this.subprocess.kill('SIGKILL')
```
Comment (hugomrdias) [inline src/ipfsd-daemon.js]: can we use the this behaviour https://github.com/sindresorhus/execa#killsignal-options here ? send SIGTERM first and SIGKILL only if it doesnt die.

**Comment by Stebalien** [comment]
Answering the question in https://github.com/ipfs/interop/pull/96#issuecomment-577617260, just send a SIGKILL. There's no real need to wait the 10s.

If someone wants to try to debug this:
* Wait for the test to hang.
* Send a SIGQUIT to go-ipfs.
* Look at the stacktrace (go-ipfs will send it to stderr on SIGQUIT).
  * Or file an issue with that stack trace.

**Comment by hugomrdias** [comment]
@Stebalien thanks, so maybe we should reduce the timeout, wait 5s for `api.stop()` to finish after that SIGKILL immediately.

**Comment by Stebalien** [comment]
SGTM!

**Comment by achingbrain** [comment]
@hugomrdias I've pushed changes to this branch that address the PR comments (I think) - can it go in?

File: src/ipfsd-daemon.js [hunk 5]
```diff
@@ -239,7 +239,32 @@ class Daemon {
...
+      }, this.opts.forceKillTimeout)
+    }
+
+    try {
+      await this.api.stop()
+    } catch (err) {
+      if (!killed) throw err // if was killed then ignore error
+
+      daemonLog.info('Daemon was killed')
```
Comment (hugomrdias) [inline src/ipfsd-daemon.js]:
```suggestion
      daemonLog.info('Daemon was force killed')
```

File: src/ipfsd-daemon.js [hunk 6]
```diff
@@ -239,7 +239,32 @@ class Daemon {
...
+    if (this.opts.forceKill !== false) {
+      killTimeout = setTimeout(() => {
+        // eslint-disable-next-line no-console
+        console.error(new Error(`Timeout stopping ${this.opts.type} node. Process ${this.subprocess.pid} will be force killed now.`))
+        killed = true
+
+        if (this.opts.type === 'go') {
+          // will make go-IPFS print a stack trace to stderr
+          this.subprocess.kill('SIGQUIT')
+        }
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply src/ipfsd-daemon.js]: Have removed it.

**Pull request merged**